org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -553,6 +555,24 @@ public synchronized void stop() {
 		return uri == null ? null : connect(uri, document);
 	}
 
+	/**
+	 * Submit an asynchronous call (i.e. to the language server) that be inserted into the current position w.r.t.
+	 * document change events
+	 *
+	 * @param <U> Computation return type
+	 * @param file Document to serialise on
+	 * @param fn Asynchronous computation on the language server
+	 * @return Asynchronous result object.
+	 */
+	<U> @Nullable CompletableFuture<VersionedResult<U>> executeOnCurrentVersionAsync(URI uri,
```

Comment:
I'm not found of a new API for that as I hope it would be guaranteed internally without requiring particular execution path. Would it be possible to change implementation of `getInitializedServer` so it returns a language server that is guaranteed to have received last document changes?

Reply:
Unfortunately I don't think you can guarantee that without blocking the UI thread completely and preventing further changes.

Let's suppose `lastChangeFuture` reflects a notification that has been sent to the server about a document change.

At this point, we compose `lastChangeFuture` with a call to the server to request formatting.

When this completes, it is guaranteed that the server knows about all the changes up to the point where we made the formatting request, so the formatting data returned will be consistent with the server's view of the document *at that point*. Unfortunately, it doesn't guarantee that the server's view of the document is up to date and no changes have been made - just that the server hasn't been told about them yet!

If we use the changes returned in the formatting request future `andThen` schedule an update on the UI thread, it is entirely possible corruption will take place, because the server's view of the document used to suggest the formatting edits is no longer accurate.

The change proposed here guards against this by allowing consuming code to detect when the document on which it wants to apply changes is inconsistent with the server's view of the document when computing those changes. Essentially this is an optimistic locking model - we can't easily prevent inconsistencies without locking threads and preventing changes. Instead we presume that mostly changes won't interfere with one another, but we can check when this is about to happen and avoid making a conflicting change (essentially a lost update).

Reply:
OK, so that brings another question: do we really want to get the initialized language server for a particular document without having all the monitored document relatively up-to-date?
Basically is there a reason to prefer current `getInitializedServer` over some newer `getUpToDateServer` ?

Reply:
I don't think `getUpToDateServer` is possible without blocking updates from happening, and unfortunately relatively-up-to-date isn't enough to be bullet-proof against possible document corruption. See below.